moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp

```diff
@@ -1547,4 +1769,217 @@ void TrajectoryExecutionManager::loadControllerParams()
+                        context.trajectory_parts_[i].multi_dof_joint_trajectory.points.back().time_from_start);
+
+      if (longest_part < 0 ||
+          std::max(context.trajectory_parts_[i].joint_trajectory.points.size(),
+                   context.trajectory_parts_[i].multi_dof_joint_trajectory.points.size()) >
+              std::max(context.trajectory_parts_[longest_part].joint_trajectory.points.size(),
+                       context.trajectory_parts_[longest_part].multi_dof_joint_trajectory.points.size()))
+        longest_part = i;
+    }
+
+    // prefer controller-specific values over global ones if defined
+    // TODO: the controller-specific parameters are static, but override
+    //       the global ones are configurable via dynamic reconfigure
+    std::map<std::string, double>::const_iterator scaling_it =
+        controller_allowed_execution_duration_scaling_.find(context.controllers_[i]);
+    const double current_scaling = scaling_it != controller_allowed_execution_duration_scaling_.end() ?
+                                       scaling_it->second :
+                                       allowed_execution_duration_scaling_;
+
+    std::map<std::string, double>::const_iterator margin_it =
+        controller_allowed_goal_duration_margin_.find(context.controllers_[i]);
+    const double current_margin =
+        margin_it != controller_allowed_goal_duration_margin_.end() ? margin_it->second : allowed_goal_duration_margin_;
+
+    // expected duration is the duration of the longest part
+    expected_trajectory_duration =
+        std::max(d * current_scaling + ros::Duration(current_margin), expected_trajectory_duration);
+  }
+
+  // construct a map from expected time to state index, for easy access to expected state location
+  if (longest_part >= 0)
+  {
+    if (context.trajectory_parts_[longest_part].joint_trajectory.points.size() >=
+        context.trajectory_parts_[longest_part].multi_dof_joint_trajectory.points.size())
+    {
+      ros::Duration d(0.0);
+      if (context.trajectory_parts_[longest_part].joint_trajectory.header.stamp > current_time)
+        d = context.trajectory_parts_[longest_part].joint_trajectory.header.stamp - current_time;
+      for (auto& point : context.trajectory_parts_[longest_part].joint_trajectory.points)
+        time_index_.push_back(current_time + d + point.time_from_start);
+    }
+    else
+    {
+      ros::Duration d(0.0);
+      if (context.trajectory_parts_[longest_part].multi_dof_joint_trajectory.header.stamp > current_time)
+        d = context.trajectory_parts_[longest_part].multi_dof_joint_trajectory.header.stamp - current_time;
+      for (auto point : context.trajectory_parts_[longest_part].multi_dof_joint_trajectory.points)
+        time_index_.push_back(current_time + d + point.time_from_start);
+    }
+  }
+
+  for (auto handle : handles)
+  {
+    if (execution_duration_monitoring_)
+    {
+      if (!handle->waitForExecution(expected_trajectory_duration))
```

Comment:
doesn't this potentially wait handles.size()*(expected_trajectory_duration - epsilon) if the first handle finishes just in time, the second takes 2 * duration and so on. I know it's contrived but ...

```suggestion
      if (!handle->waitForExecution(expected_trajectory_duration - (ros::Time::now() - current_time))
```

Reply:
Yes, this actually does not makes much sense to me either. It technically waits for the `expected trajectory duration` time for each handle